File: loot&lasers/Scripts/scale_design_to_1440.mjs
/**
 * One-time 1920→2560 design conversion: multiply authored UI sizes by 4/3.
 * Run: node "loot&lasers/scripts/scale_design_to_1440.mjs"
 *
 * After success, set ResolutionManager.USE_LEGACY_CONTENT_LAYOUT = false
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const SCALE = 4 / 3;
const SKIP = new Set([
  "ResolutionRules.gd",
  "ResolutionManager.gd",
  "scale_design_to_1440.py",
  "scale_design_to_1440.mjs",
]);

function scaleInt(n) {
  if (Math.abs(n) <= 4) return n;
  const s = Math.round(n * SCALE);
  return n > 0 ? Math.max(1, s) : Math.min(-1, s);
}

function convert(text) {
  let out = text;
  out = out.replace(
    /add_theme_font_size_override\(\s*["']font_size["']\s*,\s*(\d+)\s*\)/g,
    (_, n) => `add_theme_font_size_override("font_size", ${scaleInt(+n)})`
  );
  out = out.replace(/\.font_size\s*=\s*(\d+)/g, (_, n) => `.font_size = ${scaleInt(+n)}`);
  out = out.replace(
    /custom_minimum_size\.(x|y)\s*=\s*(\d+)/g,
    (_, axis, n) => `custom_minimum_size.${axis} = ${scaleInt(+n)}`
  );
  out = out.replace(
    /custom_minimum_size\s*=\s*Vector2(i?)\((\d+)\s*,\s*(\d+)\)/g,
    (_, i, a, b) => `custom_minimum_size = Vector2${i}(${scaleInt(+a)}, ${scaleInt(+b)})`
  );
  out = out.replace(/offset_(top|bottom|left|right)\s*=\s*(-?\d+)/g, (m, axis, n) => {
    n = +n;
    if (Math.abs(n) < 6) return m;
    return `offset_${axis} = ${scaleInt(n)}`;
  });
  return out;
}

function walk(dir, out = []) {
  for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, ent.name);
    if (ent.isDirectory()) {
      if (ent.name === ".godot" || ent.name === "scripts") continue;
      walk(p, out);
    } else if (ent.name.endsWith(".gd")) {
      out.push(p);
    }
  }
  return out;
}

const changed = [];
for (const file of walk(ROOT)) {
  const name = path.basename(file);
  if (SKIP.has(name)) continue;
  const original = fs.readFileSync(file, "utf8");
  const updated = convert(original);
  if (updated !== original) {
    fs.writeFileSync(file, updated);
    changed.push(path.relative(ROOT, file));
  }
}

console.log(`updated ${changed.length} files`);
for (const f of changed) console.log(f);
console.log("\nNext: set ResolutionManager.USE_LEGACY_CONTENT_LAYOUT = false");

File: loot&lasers/Scripts/scale_design_to_1440.py
"""One-time 1920→2560 design conversion: multiply authored UI sizes by 4/3."""
from __future__ import annotations

import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SCALE = 4.0 / 3.0
SKIP = {"ResolutionRules.gd", "ResolutionManager.gd"}

FONT_PAT = re.compile(r'(add_theme_font_size_override\(\s*["\']font_size["\']\s*,\s*)(\d+)(\s*\))')
FONT_EQ_PAT = re.compile(r"(\.font_size\s*=\s*)(\d+)")
CMIN_XY = re.compile(r"(custom_minimum_size\.(x|y)\s*=\s*)(\d+)")
CMIN_VEC = re.compile(r"(custom_minimum_size\s*=\s*Vector2i?\()(\d+)\s*,\s*(\d+)(\))")
OFFSET_PAT = re.compile(r"(offset_(?:top|bottom|left|right)\s*=\s*)(-?\d+)")


def scale_int(n: int, min_keep: int = 1) -> int:
    if abs(n) <= 4:
        return n
    s = int(round(n * SCALE))
    if n > 0:
        return max(min_keep, s)
    return min(-min_keep, s)


def convert(text: str) -> str:
    text = FONT_PAT.sub(lambda m: f"{m.group(1)}{scale_int(int(m.group(2)))}{m.group(3)}", text)
    text = FONT_EQ_PAT.sub(lambda m: f"{m.group(1)}{scale_int(int(m.group(2)))}", text)
    text = CMIN_XY.sub(lambda m: f"{m.group(1)}{scale_int(int(m.group(3)))}", text)
    text = CMIN_VEC.sub(
        lambda m: f"{m.group(1)}{scale_int(int(m.group(2)))}, {scale_int(int(m.group(3)))}{m.group(4)}",
        text,
    )

    def repl_offset(m: re.Match[str]) -> str:
        n = int(m.group(2))
        if abs(n) < 6:
            return m.group(0)
        return f"{m.group(1)}{scale_int(n)}"

    return OFFSET_PAT.sub(repl_offset, text)


def main() -> None:
    changed: list[str] = []
    for path in ROOT.rglob("*.gd"):
        if path.name in SKIP or ".godot" in path.parts:
            continue
        original = path.read_text(encoding="utf-8")
        updated = convert(original)
        if updated != original:
            path.write_text(updated, encoding="utf-8")
            changed.append(str(path.relative_to(ROOT)))
    print(f"updated {len(changed)} files")
    for name in changed:
        print(name)


if __name__ == "__main__":
    main()
